server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -605,23 +603,7 @@ protected IndexItemResponse insert(Indexer indexer,
             item.insertValues(rawIndexer.addGeneratedValues(item));
         }
 
-        Term uid = new Term(IdFieldMapper.NAME, Uid.encodeId(item.id()));
-        assert VersionType.INTERNAL.validateVersionForWrites(version);
-        Engine.Index index = new Engine.Index(
-            uid,
-            parsedDoc,
-            SequenceNumbers.UNASSIGNED_SEQ_NO,
-            indexShard.getOperationPrimaryTerm(),
-            version,
-            VersionType.INTERNAL,
-            Engine.Operation.Origin.PRIMARY,
-            startTime,
-            item.autoGeneratedTimestamp(),
-            isRetry,
-            SequenceNumbers.UNASSIGNED_SEQ_NO,
-            SequenceNumbers.UNASSIGNED_PRIMARY_TERM
-        );
-        IndexResult result = indexShard.index(index);
+        IndexResult result = indexShard.applyIndexOperationOnPrimary(parsedDoc, version, item.autoGeneratedTimestamp(), isRetry);
```

Comment:
This would exclude the indexer.index from the start time measurement, and without seeing the related test changes it's very hard to judge whether this move makes sense or not.

Reply:
I will push test changes then.

BTW links above are broken(

Reply:
ah sorry, I took the links from my unpushed local commit.

These should work:

https://github.com/crate/crate/blob/f3338443da87207f425205f88156c53d4a8b0141/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java#L776

https://github.com/crate/crate/blob/f3338443da87207f425205f88156c53d4a8b0141/server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java#L624

Reply:
> This would exclude the indexer.index from the start time measurement

I put back time computation before` indexer.index `call. Wanted to reduce number of args but no need to do that change.

Other tests needs a bigger overhaul. `IndexShardTestCase.indexDoc` uses 5.2-like source parsing approach with `SourceToParse` usage. I pushed https://github.com/crate/crate/pull/15713/commits/a0b79a7023ec6eed46d274dc7f85ff3d7e42a1e1 to show the plan to keep coverage for all those tests and use Indexer but sometimes we still need source parsing (`testTranslogRecoveryWorksWithIMC` and in general, translog code path)

Reply:
Ie to really cover all `IndexShardTestCase` subclasses, we need to use IndexItem instead of passing the source